library/alloc/src/alloc.rs

```diff
@@ -88,7 +94,10 @@ pub unsafe fn alloc(layout: Layout) -> *mut u8 {
     unsafe {
         // Make sure we don't accidentally allow omitting the allocator shim in
         // stable code until it is actually stabilized.
+        #[cfg(bootstrap)]
         core::ptr::read_volatile(&__rust_no_alloc_shim_is_unstable);
+        #[cfg(not(bootstrap))]
+        __rust_no_alloc_shim_is_unstable();
```

Comment:
Would `read_volatile(__rust_no_alloc_shim_is_unstable.addr().cast::<u8>())` (using the unstable `FnPtr::addr` method) work? Or maybe one of OpenBSD's exploit mitigations would break with that. Do we even care about that?

Reply:
That would work for most platforms.

I'm not sure about OpenBSD, but it would certainly cause issues in Xbox kernel-mode, as it loads binaries as Execute-Only (i.e., they are not readable).

Again, I suspect that the volatile read is more expensive than a tail call to empty function as it may generate load-acquire semantics.

I'm not sure if we could get away with just getting the address of the function and not reading from it? Backends might optimize that away though...

Reply:
Maybe `black_box(__rust_no_alloc_shim_is_unstable as fn())`?

Reply:
> Would read_volatile(__rust_no_alloc_shim_is_unstable.addr().cast::<u8>()) (using the unstable FnPtr::addr method) work?

Miri would definitely complain about that.

Reply:
> Maybe `black_box(__rust_no_alloc_shim_is_unstable as fn())`?

Doesn't work: we get the correct behavior where it is required for linking to work, but then the codegen tests fail as the compiler is not able to optimize away allocations.